src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/ChosenPaymentMethodEligibilityValidatorSpec.php

```diff
@@ -0,0 +1,118 @@
+    }
+
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldImplement(ConstraintValidatorInterface::class);
+    }
+
+    function it_throws_an_exception_if_constraint_is_not_an_instance_of_chosen_payment_method_eligibility(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [new ChoosePaymentMethod('PAYMENT_METHOD_CODE'), new class() extends Constraint {}])
+        ;
+    }
+
+    function it_throws_an_exception_if_value_is_not_an_instance_of_choose_payment_method_command(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [new CompleteOrder(), new ChosenShippingMethodEligibility()])
+        ;
+    }
+
+    function it_adds_violation_if_chosen_payment_method_does_not_match_supported_methods(
+        PaymentRepositoryInterface $paymentRepository,
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        PaymentMethodsResolverInterface $paymentMethodsResolver,
+        ExecutionContextInterface $executionContext,
+        PaymentMethodInterface $paymentMethod,
+        PaymentMethodInterface $differentPaymentMethod,
+        PaymentInterface $payment
+    ): void {
+        $command = new ChoosePaymentMethod('PAYMENT_METHOD_CODE');
+        $command->setOrderTokenValue('ORDER_TOKEN');
+        $command->setSubresourceId('123');
+
+        $paymentMethodRepository->findOneBy(['code' => 'PAYMENT_METHOD_CODE'])->willReturn($paymentMethod);
+        $paymentMethod->getName()->willReturn('offline');
+
+        $paymentRepository->find('123')->willReturn($payment);
+
+        $paymentMethodsResolver->getSupportedMethods($payment)->willReturn([$differentPaymentMethod]);
+
+        $executionContext
+            ->addViolation('sylius.payment_method.not_available', ['%name%' => 'offline'])
+            ->shouldBeCalled()
+        ;
+
+        $this->validate($command, new ChosenPaymentMethodEligibility());
+    }
+
+    function it_does_nothing_if_chosen_payment_method_matches_supported_methods(
+        PaymentRepositoryInterface $paymentRepository,
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        PaymentMethodsResolverInterface $paymentMethodsResolver,
+        ExecutionContextInterface $executionContext,
+        PaymentMethodInterface $paymentMethod,
+        PaymentInterface $payment
+    ): void {
+        $command = new ChoosePaymentMethod('PAYMENT_METHOD_CODE');
+        $command->setOrderTokenValue('ORDER_TOKEN');
+        $command->setSubresourceId('123');
+
+        $paymentMethodRepository->findOneBy(['code' => 'PAYMENT_METHOD_CODE'])->willReturn($paymentMethod);
+        $paymentMethod->getName()->willReturn('offline');
+
+        $paymentRepository->find('123')->willReturn($payment);
+
+        $paymentMethodsResolver->getSupportedMethods($payment)->willReturn([$paymentMethod]);
```

Comment:
Would be good to return more than one supported methods 🖖

Reply:
maybe, but IMO single one is enough :)

Reply:
Is it? Maybe the implementation checks only the first payment method from the returned array and would give us false negatives if we have more than one payment method returned? Such an implementation would work with this spec, but is not the desired functionality 😄 #mutationTestingFTW